src/tests/ftest/cart/test_group_np_srv.c

```diff
@@ -43,17 +42,13 @@ swim_crt_event_cb(d_rank_t rank, enum crt_event_source src,
 		"rank = %d, "
 		"crt_event_source = %d, "
 		"crt_event_type = %d\n",
-		 rank, src, type);
+		rank, src, type);
 
-	/* convert integer to string */
-	snprintf(type_to_a, 2, "%d", type);
+	swim_state_str[0] = type + '0';
+	swim_state_str[1] = 0;
 
-	if (swim_status_by_rank[rank] == NULL) {
-		swim_status_by_rank[rank] = (char *)malloc(max * sizeof(char));
-		strcpy(swim_status_by_rank[rank], "\0");
-	}
-
-	strcat(swim_status_by_rank[rank], type_to_a);
+	if (strlen(swim_seq_by_rank[rank]) < MAX_SWIM_STATUSES)
+		strcat(swim_seq_by_rank[rank], swim_state_str);
```

Comment:
If you are worried that string length is  greater that MAX_SWIM_STATUS, then you should use strnlen.  
If the length lf swim_seq_by_rank is between MAX_SWIM_STATUS - strlen( swim_state_str)  and MAX_SWIM_STATUS ,
then your strcat will overrun its buffer.

Reply:
> If you are worried that string length is greater that MAX_SWIM_STATUS, then you should use strnlen.

If I replace this condition:

`strlen(swim_seq_by_rank[rank]) < MAX_SWIM_STATUSES`

- wouldn't the `strnlen` version of the condition always evaluate to true?

`strnlen(swim_seq_by_rank[rank], MAX_SWIM_STATUSES) < MAX_SWIM_STATUSES`

I *want* to scan past the end of the MAX_SWIM_STATUSES to catch the overflow.

Reply:
In memory, you have string[rank=0], string[rank=1],  ... string[rank=MAX_RANK -1],  If one of your strings is of length MAX_SWIM_STATUES, then strlen is just going into the next higher rank.  strnlen will just prevent that.  You need to modify the test to be "< MAX_SWIM_STATUS - strlen(swim_state_str)".  Otherwise, when you cat the swim_state_str onto swim_seq_by_rank, you will be overwritting the seq_by_rank of the next higher rank.

Reply:
ee175cb